### PR TITLE
purifies the contract of CtInheritanceScanner, documents it and improves its error message

### PR DESCRIPTION
**Pull request by monperrus**
contract is now:
if a method visitX exists, it must be invoked
if a method scanX exists, it must be invoked

(change: we don't force to write empty visitX and scanX methods for new metamodel elements)

for example, the failure message is now:
visitCtCodeSnippetExpression does not call scanCtExpression

### CODE REVIEW

File: src/test/java/spoon/reflect/visitor/CtInheritanceScannerTest.java [hunk 1]
```diff
@@ -65,23 +73,29 @@
...
+			Method mth=null;
+			
+			// if a method visitX exists, it must be invoked
+			try {				
 				mth = CtInheritanceScanner.class.getDeclaredMethod("visit" + intf.getSimpleName(), intf);
-
-				// we should not have a scan method
-				try {
-					CtInheritanceScanner.class.getDeclaredMethod("scan" + intf.getSimpleName(), intf);
-					Assert.fail("should not have both visit" + intf.getSimpleName() + " and scan" + intf.getSimpleName()
```
Comment (GerardPaligot) [inline src/test/java/spoon/reflect/visitor/CtInheritanceScannerTest.java]: Why did you remove this assert?

Comment (GerardPaligot) [inline src/test/java/spoon/reflect/visitor/CtInheritanceScannerTest.java]: You didn't check no more if we have both visit and scan methods for one interface.

**Comment by petitpre** [comment]
I'm not agree with the contract you propose. For me:

- a visitXXX method should be define in CtVisitor for all concrete elements in model.
- a scanXXX method should be define in CtInheritanceScanner for all abstract concept in model.

so, for any new element in spoon model, you SHOULD have one visitXXX method CtVisitor if the element could be instanced, OR a scanXXX method in CtInheritanceScanner if it's an abstract concept.

**Comment by monperrus** [comment]
I don't propose a new contract, I've purified the existing one.

> you SHOULD have one visitXXX method CtVisitor if the element could be
> instanced, OR a scanXXX method in CtInheritanceScanner if it's an
> abstract concept.
I do agree with this one, and this sentence can be put verbatim in the
test case.

Can you propose a PR on my PR to add it?
